examples/cc2538-common/test-uart.c

```diff
@@ -0,0 +1,113 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * This file is part of the Contiki operating system.
+ *
+ */
+/**
+ * \addtogroup cc2538-examples
+ * @{
+ *
+ * \defgroup cc2538-test-uart Test the CC2538 UART
+ *
+ * Demonstrates the use of the CC2538 UART
+ *
+ * @{
+ *
+ * \file
+ *         A quick program for testing the UART1 on the cc2538 based
+ * platform which can be used to interface with sensor with UART
+ * interface and to communicate with other development boards
+ * \author
+ *         Akshay P M <akshaypm@ece.iisc.ernet.in>
+ */
+/*---------------------------------------------------------------------------*/
+#include "contiki.h"
+#include "sys/etimer.h"
+#include "sys/rtimer.h"
+#include "dev/leds.h"
+#include "dev/uart.h"
+#include "dev/serial-line.h"
+#include "dev/slip.h"
+#include <stdio.h>
+#include <stdint.h>
+/*---------------------------------------------------------------------------*/
+static struct etimer et;
+/*---------------------------------------------------------------------------*/
+PROCESS(cc2538_uart_demo_process, "cc2538 uart demo process");
+AUTOSTART_PROCESSES(&cc2538_uart_demo_process);
+/*---------------------------------------------------------------------------*/
+unsigned int
+uart1_send_bytes(const unsigned char *s, unsigned int len)
+{
+  unsigned int i = 0;
+
+  while(s && *s != 0) {
+    if(i >= len) {
+      break;
+    }
+    uart_write_byte(1, *s++);
+    i++;
+  }
+  return i;
+}
+/*---------------------------------------------------------------------------*/
+PROCESS_THREAD(cc2538_uart_demo_process, ev, data)
+{
+  uint16_t rxdata = 0;
+  static uint16_t count;
+  char string[count];
+  extern process_event_t serial_line_event_message;
+  
+  PROCESS_BEGIN();
+  uart_set_input(1, serial_line_input_byte);
+  etimer_set(&et, CLOCK_SECOND * 7);
+  leds_toggle(LEDS_GREEN);
+
+  while(1) {
+    PROCESS_WAIT_EVENT_UNTIL(etimer_expired(&et));
+    sprintf(string, "sending string\n", ++count);
+    uart1_send_bytes((uint8_t *)string, sizeof(string) - 1);
+    etimer_reset(&et);
+    printf("string sent\n");
+
+    PROCESS_WAIT_UNTIL(etimer_expired(&et));
```

Comment:
I meant that `PROCESS_WAIT_UNTIL()` should replace the `PROCESS_WAIT_EVENT_UNTIL()` in line 93.  Both of those macroes do two things:

- a test
- yield to other processes

The macro with "`_EVENT`" in its name yields first.  The macro without "`_EVENT`" in its name tests first.  Doing the _test_ first allows this process "to catch up" to the etimer -- if necessary.